src/engine.rs

```diff
@@ -6,14 +6,38 @@ use crate::interpreter::*;
 use crate::lexer::*;
 use crate::parser::*;
 use crate::scheduler::*;
-use crate::utils::gather_functions;
+use crate::utils::{gather_functions, UtilsError};
 use crate::value::*;
 use crate::{Extension, QueryResults};
 
+use thiserror::Error;
+
 use std::convert::AsRef;
 use std::path::Path;
 
-use anyhow::{bail, Result};
+type Result<T> = std::result::Result<T, EngineError>;
+
+#[derive(Error, Debug)]
+pub enum EngineError {
+    #[error("interpreter error: {0}")]
+    InterpreterError(#[from] InterpreterError),
+    #[error("lexer error: {0}")]
+    LexerError(#[from] LexerError),
+    #[error("parser error: {0}")]
+    ParserError(#[from] ParserError),
+    #[error("scheduler error: {0}")]
+    SchedulerError(#[from] SchedulerError),
+    #[error("utils error: {0}")]
+    UtilsError(#[from] UtilsError),
+    #[error("value error: {0}")]
+    ValueError(#[from] ValueError),
+    #[error("data must be object")]
+    DataMustBeObject,
```

Comment:
Say in a future version we refactor the code and eliminate the chance of such an error, would that be a breaking change?